typed-racket-lib/typed-racket/utils/opaque-object.rkt

```diff
@@ -57,21 +60,54 @@
       (dynamic-object/c (append methods remaining-methods)
                         (append method-ctcs
                                 (for/list ([m remaining-methods])
-                                  (restrict-typed->/c)))
+                                  (restrict-typed->/c m)))
                         (append fields remaining-fields)
                         (append field-ctcs
                                 (for/list ([m remaining-fields])
-                                  (restrict-typed-field/c obj m)))))
+                                  (restrict-typed-field/c m)))))
     ;; FIXME: this is a bit sketchy because we have to construct
     ;;        a contract that depends on the actual object that we got
     ;;        since we don't know its methods beforehand
     (((contract-late-neg-projection guard/c) blame) obj neg-party)))
 
+(define (object/c-opaque-name ctc)
+  (build-object/c-type-name 'object/c-opaque
+                            (base-object/c-opaque-method-names ctc)
+                            (base-object/c-opaque-method-ctcs ctc)
+                            (base-object/c-opaque-field-names ctc)
+                            (base-object/c-opaque-field-ctcs ctc)))
+
+;; Similar to object/c-stronger, but without width subtyping.
+;; (Intuition: unspecified fields are guarded by the strongest possible contract)
+;; An opaque object contract `this` is stronger than `that` when:
+;; - `that` is an opaque contract
+;;   and `this` specifies at most the same members as `that`
+;;   and `this` has stronger contracts on all members
+;; - `that` is an object/c contract
+;;   and `this` has stronger contracts on their common members
+(define (object/c-opaque-stronger this that)
+  (define that-opaque? (base-object/c-opaque? that))
+  (cond
+   [(or that-opaque?
+        (base-object/c? that))
+    (define this-ctc (base-object/c-opaque-obj/c this))
+    (define that-ctc (if that-opaque? (base-object/c-opaque-obj/c that) that))
+    (and
+      (object/c-common-methods-stronger? this-ctc that-ctc)
+      (object/c-common-fields-stronger? this-ctc that-ctc)
+      (when that-opaque?
```

Comment:
I'd write `or` not `when` here. More generally, nothing really "smells" wrong here. I'd say that if you have a good set of tests, then this is probably right.